### ISSUE TITLE
No binary distributions for Microsoft Windows.

### ISSUE THREAD

**Issue by Ri0n** [cyrusimap/cyrus-sasl#506]
Since it's quite hard to build cyrus-sasl for Windows, it's better its developers take care of this.

I'd expect ready to use downloads
1) debug/release builds of cyrus sasl
2) x86/x64 builds
3) static and dlls (static plugins would be very good too to reduce app load time)
4) Visual Studio 2015/2017 support.
5) distribution with headers and anything else required for development.
6) mingw-w64 builds would be useful too (but not for me)

From the above I'd very appreciate having x64 static libraries compiled with VS2015.

I'm ready to help porting everything to MSBuild, but need some explanation of status of NTMakefiles since they look outdated/untested.

PS I didn't read yes Issue #499

**Comment by sfhacker** [comment]
Hundreds of people and companies have been using Windows binary distributions for some time now!

Have you done a quick Google search?

We are more than happy to assist you with your building and/or compilation issues.

**Comment by Ri0n** [comment]
yes. google found nothing useful except Issue #499 which basically says it's broken now. And I also tried to use those NTMakefiles but got just errors. ( I fixed some but not all ).

First of all about compilation is why plugins dir has db_berkeley while the same code is sasldb? Can I just disable db_berkeley plugin and make a pull request with this change? 

In other words I'd like to have NTmakefiles working first to understand what exactly and how should be compiled for Windows. Then it's not a problem to implement vcxproj or whatever.

**Comment by Ri0n** [comment]
And of course it's better if developers setup nightly windows builds to provide binary snapshots and control code quality (ideally buildable w/o warnings).

**Comment by sfhacker** [comment]
> yes. google found nothing useful except Issue #499 which basically says it's broken now.
This is most interesting! The organisation I work for has been using a Windows binary distribution since 2011.

We haven't used Berkeley DB for some time now. I've read somewhere that it should be avoided.

If you either post your building/compilation issues here or email them to us, we are more than happy to provide some assistance.

**Comment by Ri0n** [comment]
It seems Windows code wasn't updated for more than a decade. It even still missed unicode support.

**Comment by Ri0n** [comment]
even so I was able to build core sasl library (no plugins yet) with vcxproj files. Most likely it will crash or just won't work though because of the mess with unicode.

[sasl_debug.zip](https://github.com/cyrusimap/cyrus-sasl/files/1850644/sasl_debug.zip)

**Comment by Ri0n** [comment]
I made a pull request with basic stuff https://github.com/cyrusimap/cyrus-sasl/pull/507

**Comment by sfhacker** [comment]
What are we supposed to do with 'sasl_debug.zip'?
You need/want us to test it or ...?

**Comment by Ri0n** [comment]
ignore sasl_debug.zip, I've fixed a lot of stuff after it.
The pull request compiles w/o warnings and with some unicode support (probably incomplete, since I didn't track all the code yet)

And I was unable to find any binaries compiled with Visual Studio. I need Visual Studio because I don't want to mix toolchains and their runtimes etc. And I need sasl2 static or dynamic library to link it with corresponding [qca](https://github.com/KDE/qca/blob/master/plugins/qca-cyrus-sasl/) plugin.

**Comment by Ri0n** [comment]
It's interesting what to do with static plugins.
I think to add another vcxproj which will link static sasl2 and static plugins together.

**Comment by Ri0n** [comment]
It seems my pull request has reache a point where it's enough for my needs except a few minor details.
I'll test compiled dlls and let you know if something goes wrong.

**Comment by Ri0n** [comment]
I fixed a few more problems after and updated pull request. In general test results are good enough.

I think cyrus-sasl developers can setup AppVeyor or other similar tools to get Windows binaries for free and make them available for community.

**Comment by Ri0n** [comment]
I finally found some time to put dependencies on Conan too.
Now it's sasl turn.
